overlord/state/copy.go

```diff
@@ -0,0 +1,122 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2020 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package state
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/snapcore/snapd/jsonutil"
+	"github.com/snapcore/snapd/osutil"
+)
+
+type checkpointOnlyBackend struct {
+	path string
+}
+
+func (b *checkpointOnlyBackend) Checkpoint(data []byte) error {
+	return osutil.AtomicWriteFile(b.path, data, 0600, 0)
+}
+
+func (b *checkpointOnlyBackend) EnsureBefore(d time.Duration) {
+	panic("cannot use EnsureBefore in checkpointOnlyBackend")
+}
+
+func (b *checkpointOnlyBackend) RequestRestart(t RestartType) {
+	panic("cannot use RequestRestart in checkpointOnlyBackend")
+}
+
+func copyData(subkeys []string, pos int, srcData map[string]*json.RawMessage, dstData map[string]interface{}) error {
+	raw, ok := srcData[subkeys[pos]]
+	if !ok {
+		return ErrNoState
+	}
+
+	if pos+1 == len(subkeys) {
+		dstData[subkeys[pos]] = raw
+		return nil
+	}
+
+	var srcDatam map[string]*json.RawMessage
+	if err := jsonutil.DecodeWithNumber(bytes.NewReader(*raw), &srcDatam); err != nil {
+		return fmt.Errorf("option %q is not a map", strings.Join(subkeys[:pos+1], "."))
```

Comment:
Should you include `err` in the error message?

Reply:
+1, I think the error could just say "cannot unmarshal option %q: %v".

Reply:
these are not options though,  s/option/state entry/. maybe

"cannot unmarshal state entry %q as a map: ..."

Reply:
this scenario, basically scalar where we expect nesting instead, probably merits a test